src/contrib/cluster/Akka.Cluster.Tools/Client/ClusterClient.cs

```diff
@@ -373,12 +466,138 @@ protected override void PostStop()
+    }
+    
+    private bool Discovering(object message)
+    {
+        switch (message)
+        {
+            case DiscoverTick:
+                Rediscover();
+                return true;
+            
+            case ServiceDiscovery.Resolved resolved:
+                _discoveryCancelable?.Cancel();
+                _discoveryCancelable = null;
+                
+                if (resolved.Addresses.Count == 0)
+                {
+                    // discovery didn't find any contacts, retry discovery
+                    _discoveryCancelable = Context.System.Scheduler.ScheduleTellOnceCancelable(
+                        delay: _settings.DiscoverySettings.DiscoveryRetryInterval, 
+                        receiver: Self,
+                        message: DiscoverTick.Instance,
+                        sender: Self);
+                    return true;
+                }
+
+                _contacts = resolved.Addresses.Select(address => {
+                    var path = ResolvedTargetToReceptionistActorPath(address);
+                    return new Contact(path, Context.ActorSelection(path));
+                }).ToImmutableHashSet();
+
+                VerifyContacts().PipeTo(Self, Self);
+                
+                async Task<ResolveResult[]> VerifyContacts()
+                {
+                    var tasks = _contacts.Select(c => ResolveContact(c, TimeSpan.FromSeconds(1), default));
+                    return await Task.WhenAll(tasks);
+                }
+                
+                return true;
+            
+            case ResolveResult[] resolved:
+                _contacts = resolved.Where(r => r.Subject is not null).Select(r => r.Contact).ToImmutableHashSet();
+                if (_contacts.Count == 0)
+                {
+                    _log.Warning("Cluster.Client contact point resolution phase failed, will try again.");
+                    _discoveryCancelable = Context.System.Scheduler.ScheduleTellOnceCancelable(
+                        delay: _settings.DiscoverySettings.DiscoveryRetryInterval, 
+                        receiver: Self,
+                        message: DiscoverTick.Instance,
+                        sender: Self);
+                }
+                else
+                {
+                    Reestablish();
+                }
+                
+                return true;
+            
+            case DiscoveryFailure fail:
+                _log.Warning(fail.Cause, "Cluster.Client contact point service discovery phase failed, will try again.");
+                _discoveryCancelable = Context.System.Scheduler.ScheduleTellOnceCancelable(
+                    delay: _settings.DiscoverySettings.DiscoveryRetryInterval, 
+                    receiver: Self,
+                    message: DiscoverTick.Instance,
+                    sender: Self);
+                return true;
+            
+            case ClusterReceptionist.Contacts:
+                // ignored, we're re-discovering
```

Comment:
Why would we ignore this - does this message contain any valid contacts?

Reply:
I guess we could incorporate the result with the possible future discovery result, but this is risky because we're already inside the discovery phase. If there's suddenly a `Contacts` message coming when we're in this state, that means something is seriously wrong with the network / configuration and there could be a possible split-brain happening

Reply:
We don't care about split brains in this scenario though - the client just needs 1 node to talk to.

Reply:
As long as someone is answering the phone, that's good enough.